contracts/stake-cw20-external-rewards/src/contract.rs

```diff
@@ -0,0 +1,267 @@
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> Result<Response<Empty>, ContractError> {
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    let admin = match msg.admin {
+        Some(admin) => Some(deps.api.addr_validate(admin.as_str())?),
+        None => None,
+    };
+
+    let config = Config {
+        admin,
+        staking_contract: msg.staking_contract,
+        reward_token: msg.reward_token,
+    };
+    CONFIG.save(deps.storage, &config)?;
+    
+    let reward_config = RewardConfig{
+        periodFinish: 0,
+        rewardRate: Default::default(),
+        rewardDuration: 100000
+    };
+    REWARD_CONFIG.save(deps.storage, &reward_config);
+
+    Ok(Response::new())
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn execute(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> Result<Response<Empty>, ContractError> {
+    match msg {
+        ExecuteMsg::StakeChangeHook(msg) => execute_stake_changed(deps, env, info, msg),
+        ExecuteMsg::Claim {} => execute_claim(deps, env, info),
+        ExecuteMsg::Fund {} => execute_fund_native(deps, env, info),
+        ExecuteMsg::Receive(msg) => execute_receive(deps, env, info, msg)
+    }
+}
+
+pub fn execute_receive(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    wrapper: Cw20ReceiveMsg
+) -> Result<Response<Empty>, ContractError> {
+    let msg: ReceiveMsg = from_binary(&wrapper.msg)?;
+    let config = CONFIG.load(deps.storage)?;
+    let sender = deps.api.addr_validate(&*wrapper.sender)?;
+    if config.reward_token != Denom::Cw20(info.sender) {
+        return Err(Unauthorized {})
+    };
+    if config.admin != Some(sender.clone()){
+        return Err(Unauthorized {})
+    };
+    match msg {
+        ReceiveMsg::Fund { .. } => {execute_fund(deps, env, sender, wrapper.amount)}
+    }
+}
+
+pub fn execute_fund_native (
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo
+) -> Result<Response<Empty>, ContractError> {
+    let config = CONFIG.load(deps.storage)?;
+    if config.admin != Some(info.sender.clone()){
+        return Err(Unauthorized {})
+    };
+    let coin = info.funds.first()?;
+    let amount = coin.clone().amount;
+    let denom = coin.clone().denom;
+    if config.reward_token != Denom::Native(denom) {
+        return Err(Unauthorized {})
+    };
+    execute_fund(deps, env, info.sender, amount)
+}
+
+pub fn execute_fund (
+    mut deps: DepsMut,
+    env: Env,
+    sender: Addr,
+    amount: Uint128
+) -> Result<Response<Empty>, ContractError> {
+    update_rewards(&mut deps, &env, &sender);
+    
+    let reward_config = REWARD_CONFIG.load(deps.storage)?;
+    let new_reward_config = if reward_config.periodFinish < env.block.height {
```

Comment:
If I execute fund a second time before the initial rewards finish is the intended behavior to overwrite that previous reward rate?

Reply:
is that what happens here? intended behavior is to extend the current rewards period with the same rate

Reply:
This line below seems like it isn't taking into account the old rate and is just setting the rate based on the amount sent:
```rust
rewardRate: amount / reward_config.rewardDuration
```

Reply:
ok, yeah think I see what you are saying here. Maybe need to look at the tests as well, will get back to you.

Reply:
ok, here is what is happening, needed to refresh context a bit.

1. If rewards are ended, ie `reward_period_finish` < `current_block_height`, we will start a new rewards period with `amount` distributed evenly over `rewards_duration` number of blocks.

2. If rewards are ongoing we will increase the remaining rewards by `amount` evenly distributed over the remaining blocks in the reward period.

I think confusion here is `reward_config.periodFinish < env.block.height == true` means rewards are currently not ongoing so the reward amount should be overwritten.

Reply:
This is how the synthetix reward contract functions https://github.com/Synthetixio/synthetix/blob/45fd1febc8595057ea8d157730b36111c775d177/contracts/StakingRewards.sol#L113

Interested if you think a different approach is better suited.